### PR TITLE
arm64: Support for FPU profile with procfs

### PR DESCRIPTION
**Pull request by qinwei2004**
## Summary

   To reduce the count of FPU context switching will result at a
performance improve with system. but it need to balance the using of FPU instruction and register
   the patch submit a base method to see performance counts for the FPU.
   Please read README.txt at chapter of FPU Support and Performance
for more information

## Impact
None

## Testing
1. qemu-armv8a(nsh/nsh_smp)
1. fvp-armv8r(nsh/nsh_smp)

### CODE REVIEW

File: arch/arm64/src/common/arm64_fpu.c [hunk 1]
```diff
@@ -54,6 +80,45 @@
 static struct fpu_reg g_idle_thread_fpu[CONFIG_SMP_NCPUS];
 static struct arm64_cpu_fpu_context g_cpu_fpu_ctx[CONFIG_SMP_NCPUS];
 
+#ifdef CONFIG_FS_PROCFS_REGISTER
+
+/* procfs methods */
+
+static int arm64_fpu_procfs_open(struct file *filep, const char *relpath,
+                             int oflags, mode_t mode);
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:88]:
```suggestion
static int arm64_fpu_procfs_open(struct file *filep, const char *relpath,
                                 int oflags, mode_t mode);
```

File: arch/arm64/src/common/arm64_fpu.c [hunk 2]
```diff
@@ -54,6 +80,45 @@
...
+static int arm64_fpu_procfs_close(struct file *filep);
+static ssize_t arm64_fpu_procfs_read(struct file *filep, char *buffer,
+                                 size_t buflen);
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:91]:
```suggestion
static ssize_t arm64_fpu_procfs_read(struct file *filep, char *buffer,
                                     size_t buflen);
```

File: arch/arm64/src/common/arm64_fpu.c [hunk 3]
```diff
@@ -54,6 +80,45 @@
...
+  arm64_fpu_procfs_read,  /* read */
+  NULL,                   /* write */
+  NULL,                   /* dup */
+  NULL,                   /* opendir */
+  NULL,                   /* closedir */
+  NULL,                   /* readdir */
+  NULL,                   /* rewinddir */
+
+  arm64_fpu_procfs_stat   /* stat */
+};
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:112]: I think we can use designated initialiser here and remove a couple of explicitly NULL-ed items

File: arch/arm64/src/common/arm64_fpu.c [hunk 4]
```diff
@@ -84,6 +149,120 @@ static void arm64_fpu_access_trap_disable(void)
   ARM64_ISB();
 }
 
+#ifdef CONFIG_FS_PROCFS_REGISTER
+
+static int arm64_fpu_procfs_open(struct file *filep, const char *relpath,
+                             int oflags, mode_t mode)
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:155]:
```suggestion
static int arm64_fpu_procfs_open(struct file *filep, const char *relpath,
                                 int oflags, mode_t mode)
```

File: arch/arm64/src/common/arm64_fpu.c [hunk 5]
```diff
@@ -84,6 +149,120 @@ static void arm64_fpu_access_trap_disable(void)
...
+
+  /* Release the file attributes structure */
+
+  kmm_free(priv);
+  filep->f_priv = NULL;
+  return OK;
+}
+
+static ssize_t arm64_fpu_procfs_read(struct file *filep, char *buffer,
+                                 size_t buflen)
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:208]:
```suggestion
static ssize_t arm64_fpu_procfs_read(struct file *filep, char *buffer,
                                     size_t buflen)
```

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
Excellent! Thanks for including the explanation in the commit message!

File: arch/arm64/src/common/arm64_fpu.c (hunk 3 above)
Comment (qinwei2004) [reply arch/arm64/src/common/arm64_fpu.c:112]: you mean like :
const struct file_operations relay_file_operations = {
	.open		= relay_file_open,
	.poll		= relay_file_poll,
	.mmap		= relay_file_mmap,
	.read		= relay_file_read,
	.llseek		= no_llseek,
	.release	= relay_file_release,
	.splice_read	= relay_file_splice_read,
};
but most of NuttX driver use the style like my define, I guess it's for compiler compatible??

**Comment by xiaoxiang781216** [comment]
@acassis could you merge this patch?

File: arch/arm64/src/common/arm64_fpu.c [hunk 6]
```diff
@@ -84,6 +148,120 @@ static void arm64_fpu_access_trap_disable(void)
...
+    {
+      uerr("ERROR: Only O_RDONLY supported\n");
+      return -EACCES;
+    }
+
+  /* Allocate the open file structure */
+
+  priv = (struct arm64_fpu_procfs_file_s *)kmm_zalloc(
+    sizeof(struct arm64_fpu_procfs_file_s));
+  if (!priv)
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:176]:
```suggestion
  if (priv == NULL)
```

File: arch/arm64/src/common/arm64_fpu.c [hunk 7]
```diff
@@ -84,6 +148,120 @@ static void arm64_fpu_access_trap_disable(void)
...
+                                     size_t buflen)
+{
+  struct arm64_fpu_procfs_file_s *attr;
+  struct arm64_cpu_fpu_context *ctx;
+  off_t offset;
+  int linesize;
+  int ret;
+  int i;
+
+  uinfo("buffer=%p buflen=%lu\n", buffer, (unsigned long)buflen);
```
Comment (pkarashchenko) [inline arch/arm64/src/common/arm64_fpu.c:216]:
```suggestion
  uinfo("buffer=%p buflen=%zu\n", buffer, buflen);
```

**Comment by qinwei2004** [comment]
@acassis @xiaoxiang781216 
Any change is needed?? or we can merge this PR

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@qinwei2004 

I noticed that ostest crashes with qemu-armv8a:nsh_smp 


```
$ /home/ishikawa/opensource/QEMU/qemu-7.1/build/qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic -machine virt,virtualization=on,gic-version=3 -net none -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel ./nuttx
...

NuttShell (NSH) NuttX-12.1.0
nsh> uname -a
NuttX 12.1.0 4be499a243 Apr 22 2023 18:41:21 arm64 qemu-armv8a
nsh> ps
  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK           STACK   USED  FILLED COMMAND
    0     0   0   0 FIFO     Kthread N-- Assigned           0000000000000000 016336 001040   6.3%  CPU0 IDLE
    1     1   1   0 FIFO     Kthread N-- Running            0000000000000000 016336 000368   2.2%  CPU1 IDLE
    2     2   2   0 FIFO     Kthread N-- Running            0000000000000000 016336 000128   0.7%  CPU2 IDLE
    3     3   3   0 FIFO     Kthread N-- Running            0000000000000000 016336 000128   0.7%  CPU3 IDLE
    4     4 --- 192 RR       Kthread --- Waiting  Semaphore 0000000000000000 016288 001040   6.3%  hpwork 0x402c3f70
    5     5   0 100 RR       Task    --- Running            0000000000000000 016304 002784  17.0%  nsh_main
nsh> free
                   total       used       free    largest  nused  nfree
        Umem:  131210576     571152  130639424  130639328     71      2
nsh> ostest
...

user_main: semaphore test
sem_test: Initializing semaphore to 0
sem_test: Starting waiter thread 1
sem_test: Set thread 1 priority to 191
sem_test: Starting waiter thread 2
waiter_func: Thread 1 Started
sem_test: Set thread 2 priority to 128
waiter_func: Thread 1 initial semaphore value = 0
waiter_func: Thread 1 waiting on semaphore
sem_test: Starting poster thread 3
waiter_func: Thread 2 Started
sem_test: Set thread 3 priority to 64
waiter_func: Thread 2 initial semaphore value = -1
waiter_func: Thread 2 waiting on semaphore
poster_func: Thread 3 started
poster_func: Thread 3 semaphore value = -2
poster_func: Thread 3 posting semaphore
poster_func: Thread 3 new semaphore value = -1
waiter_func: Thread 1 awakened
poster_func: Thread 3 semaphore value = -1
poster_func: Thread 3 posting semaphore
waiter_func: Thread 1 new semaphore value = -1
waiter_func: Thread 2 awakened
waiter_func: Thread 1 done
waiter_func: Thread 2 new semaphore value = 0
poster_func: Thread 3 new semaphore value = 0
[CPU3] _assert: Current Version: NuttX  12.1.0 4be499a243 Apr 22 2023 18:41:21 arm64
[CPU3] _assert: Assertion failed panic: at file: common/arm64_fatal.c:375 task(CPU3): ostest 0x402a4794
[CPU3] up_dump_register: stack = 0x40389d10
[CPU3] up_dump_register: x0:   0xdeaddeaddeaddead  x1:   0x7
[CPU3] up_dump_register: x2:   0x0                 x3:   0x402c3ca0
[CPU3] up_dump_register: x4:   0x1                 x5:   0x2
[CPU3] up_dump_register: x6:   0x0                 x7:   0xa00
[CPU3] up_dump_register: x8:   0x88                x9:   0x0
[CPU3] up_dump_register: x10:  0x0                 x11:  0x0
[CPU3] up_dump_register: x12:  0x0                 x13:  0x0
[CPU3] up_dump_register: x14:  0x0                 x15:  0x0
[CPU3] up_dump_register: x16:  0x4028ca18          x17:  0x0
[CPU3] up_dump_register: x18:  0x0                 x19:  0x402c2098
[CPU3] up_dump_register: x20:  0x4038a010          x21:  0x402b5c88
[CPU3] up_dump_register: x22:  0x0                 x23:  0x0
[CPU3] up_dump_register: x24:  0x0                 x25:  0x0
[CPU3] up_dump_register: x26:  0x0                 x27:  0x0
[CPU3] up_dump_register: x28:  0x0                 x29:  0x0
[CPU3] up_dump_register: x30:  0x40282230        
[CPU3] up_dump_register: 
[CPU3] up_dump_register: STATUS Registers:
[CPU3] up_dump_register: SPSR:      0x600003c5        
[CPU3] up_dump_register: ELR:       0x4028238c        
[CPU3] up_dump_register: SP_EL0:    0x4038a040        
[CPU3] up_dump_register: SP_ELX:    0x40389d10        
[CPU3] up_dump_register: TPIDR_EL0: 0x403726f0        
[CPU3] up_dump_register: TPIDR_EL1: 0x403726f0        
[CPU3] up_dump_register: EXE_DEPTH: 0x1               
[CPU3] show_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK  STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[CPU3] show_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- -------- 0x402cc000      8192      8192   100.0%!   irq
[CPU3] dump_task:       0     0     0   0 FIFO     Kthread N-- Running            0000000000000000 0x402ce030     16336      1168     7.1%    CPU0 IDLE
[CPU3] dump_task:       1     1     1   0 FIFO     Kthread N-- Running            0000000000000000 0x402d2030     16336       368     2.2%    CPU1 IDLE
[CPU3] dump_task:       2     2     2   0 FIFO     Kthread N-- Running            0000000000000000 0x402d6030     16336       368     2.2%    CPU2 IDLE
[CPU3] dump_task:       3     3     3   0 FIFO     Kthread N-- Assigned           0000000000000000 0x402da030     16336       368     2.2%    CPU3 IDLE
[CPU3] dump_task:       4     4     0 192 RR       Kthread --- Waiting Semaphore  0000000000000000 0x402e0b00     16288      1264     7.7%    hpwork 0x402c3f70
[CPU3] dump_task:       5     5     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x402e5140     16304      2096    12.8%    nsh_main
[CPU3] dump_task:      22    22     1 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x4036e4e0     16256      2160    13.2%    ostest Arg1 Arg2 Arg3 Arg4
[CPU3] dump_task:     122    22     3 128 RR       pthread --- Running            0000000000000000 0x403863b0     16336      3680    22.5%    ostest 0x402a4794 0x2
[CPU3] dump_task:     123    22     2  64 RR       pthread --- Waiting Semaphore  0000000000000000 0x4038a3c0     16336      1808    11.0%    ostest 0x402a4608 0x3
[CPU3] dump_task:      15    15     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x40369db0     16304      1920    11.7%    ostest

```

**Comment by xiaoxiang781216** [comment]
@qinwei2004 will take a look tomorrow.

**Comment by qinwei2004** [comment]
@masayuki2009
I cannot reproduce with ubuntu 22.04.2 LTS and the default qemu  6.2.0
Do you build qemu 7.1 and reproduce this?? 
So could you give me your build configure with qemu? I want to build qemu like your and reproduce it

thanks

**Comment by masayuki2009** [comment]
@qinwei2004 

>I cannot reproduce with ubuntu 22.04.2 LTS and the default qemu 6.2.0
>Do you build qemu 7.1 and reproduce this??
>So could you give me your build configure with qemu? I want to build qemu like your and reproduce it

Yes, I build qemu-7.1 on ubuntu 20.04.6 LTS (x86_64).
In the build.log, I can see

```
# Configured with: '/home/ishikawa/opensource/QEMU/qemu-7.1/configure' '--target-list=riscv64-softmmu,riscv32-softmmu,aarch64-softmmu,arm-softmmu,i386-softmmu,x86_64-softmmu' '--enable-kvm'
```

Also, I've been using the following arm gcc to compile NuttX

```
$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (Arm GNU Toolchain 12.2.MPACBTI-Rel1 (Build arm-12-mpacbti.34)) 12.2.1 20230214
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by qinwei2004** [comment]
@masayuki2009 @xiaoxiang781216 
Sorry for lately reply
I check this issue, for SMP, the CONFIG_ARCH_FPU shouldn't enabled, but with configure, I can see
CONFIG_ARCH_FPU=y
the FPU context switch for FPU is little sophisticated and need to be support
for the present, you need to disable the option in boards/arm64/qemu/qemu-armv8a/configs/nsh_smp/defconfig
CONFIG_ARCH_FPU=n 
I check with myself qemu(6.2.0 and 7.2.2) and not reproduce the issue again.
please check with your environment, if you confirm the issue is fix, I will push a PR to fix it

**Comment by masayuki2009** [comment]
@qinwei2004 

I confirmed that the following change can avoid the crash on qemu-7.1

```
--- a/boards/arm64/qemu/qemu-armv8a/configs/nsh_smp/defconfig
+++ b/boards/arm64/qemu/qemu-armv8a/configs/nsh_smp/defconfig
@@ -5,6 +5,7 @@
 # You can then do "make savedefconfig" to generate a new defconfig file that includes your
 # modifications.
 #
+# CONFIG_ARCH_FPU is not set
 CONFIG_ARCH="arm64"
 CONFIG_ARCH_ARM64=y
 CONFIG_ARCH_BOARD="qemu-armv8a"
```

**Comment by qinwei2004** [comment]
@masayuki2009 
Thank for your suggestion, it seem fvp for armv8r has seem question
I will submit a PR to fix this issue.

@zyfeier

**Pull request merged**